// @json-formost/core/esm2015/lib/controls/form-control.js
import { FormControl as NgFormControl, Validators } from '@angular/forms';
import { populateInterfaceProperties } from '../util';
export class FormControl extends NgFormControl {
    constructor(source, formState) {
        super(formState, null, null);
        // type: string
        this.minLength = null;
        this.maxLength = null;
        this.pattern = null;
        this.contentMediaType = null;
        // type: number
        this.multipleOf = null;
        this.minimum = null;
        this.exclusiveMinimum = null;
        this.maximum = null;
        this.exclusiveMaximum = null;
        populateInterfaceProperties(this, source, ['type', 'required', 'enum', 'minLength', 'maxLength', 'pattern', 'format', 'contentEncoding', 'contentMediaType', 'multipleOf', 'minimum', 'exclusiveMinimum', 'maximum', 'exclusiveMaximum']);
        this.setValidators(this.getValidators());
    }
    get uiType() {
        if (this.type == 'string' && this.enum) {
            return 'dropdown';
        }
        if (this.type == 'boolean') {
            return 'checkbox';
        }
        // if (this.type == 'string' && (this.minLength > 99 || this.maxLength > 99)) { return 'bigtext'; }
        if (this.type == 'string' && this.contentMediaType == 'text/html') {
            return 'richtext';
        }
        if (this.type == 'string' && this.contentMediaType == 'text/markdown') {
            return 'markdown';
        }
        return 'input';
    }
    ;
    get inputType() {
        if (this.uiType == 'input') {
            if (this.type == 'number' || this.type == 'integer') {
                return 'number';
            }
            if (this.type == 'string' && this.format == 'date-time') {
                return 'datetime-local';
            }
            if (this.type == 'string' && this.format == 'date') {
                return 'date';
            }
            if (this.type == 'string' && this.format == 'time') {
                return 'number';
            }
            if (this.type == 'string' && this.format == 'email') {
                return 'email';
            }
            if (this.type == 'string' && this.format == 'uri') {
                return 'url';
            }
            if (this.type == 'string') {
                return 'text';
            }
        }
        return null;
    }
    populate(source) {
        populateInterfaceProperties(this, source, ['type', 'required', 'enum', 'minLength', 'maxLength', 'pattern', 'format', 'contentEncoding', 'contentMediaType', 'multipleOf', 'minimum', 'exclusiveMinimum', 'maximum', 'exclusiveMaximum']);
    }
    getControlType() {
        return 'control';
    }
    getValidators() {
        const validators = [];
        if (this.required)
            validators.push(Validators.required);
        if (this.type == 'string') {
            if (this.minLength !== null)
                validators.push(Validators.minLength(this.minLength));
            if (this.maxLength !== null)
                validators.push(Validators.minLength(this.maxLength));
            if (this.pattern !== null)
                validators.push(Validators.pattern(this.pattern));
            if (this.format == "email")
                validators.push(Validators.email);
        }
        if (this.type == 'number') {
            if (this.minimum !== null)
                validators.push(Validators.min(this.minimum));
            if (this.maximum !== null)
                validators.push(Validators.max(this.maximum));
        }
        console.info(name, validators);
        return validators;
    }
}
