// @json-formost/core/esm2015/lib/controls/form-array.js
import { FormArray as NgFormArray } from '@angular/forms';
import { populateInterfaceProperties } from '../util';
export class FormArray extends NgFormArray {
    constructor(controls, validatorOrOpts, asyncValidator, fbs) {
        super(controls, validatorOrOpts, asyncValidator);
        this.controls = controls;
        this.fbs = fbs;
        this.itemTypes = [];
        this.additionalItems = false;
    }
    populate(source) {
        populateInterfaceProperties(this, source, ['uniqueItems', 'contains', 'additionalItems', 'minItems', 'maxItems']);
    }
    getControlType() {
        return 'array';
    }
    setValue(value, options = {}) {
        if (value && value.length > 0) {
            if (!this.fbs)
                throw 'Must supply FormBuilderService instance.';
            value.forEach(val => {
                const schema = this.itemTypes.length == 1
                    ? this.itemTypes[0]
                    : this.itemTypes.find(sch => sch.type == typeof val); // ToDo: better type matching
                // console.info("schema+val", schema, val);
                let ctl = this.fbs.getFormostAbstractControl(schema);
                this.push(ctl);
            });
        }
        super.setValue(value, options);
        // this._checkAllValuesPresent(value);
        // value.forEach((newValue: any, index: number) => {
        //   this._throwIfControlMissing(index);
        //   this.at(index).setValue(newValue, {onlySelf: true, emitEvent: options.emitEvent});
        // });
        // this.updateValueAndValidity(options);
    }
}
