// @json-formost/core/bundles/json-formost-core.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('rxjs'), require('@angular/forms')) :
    typeof define === 'function' && define.amd ? define('@json-formost/core', ['exports', '@angular/core', 'rxjs', '@angular/forms'], factory) :
    (global = global || self, factory((global['json-formost'] = global['json-formost'] || {}, global['json-formost'].core = {}), global.ng.core, global.rxjs, global.ng.forms));
}(this, (function (exports, i0, rxjs, forms) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (b.hasOwnProperty(p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, exports) {
        for (var p in m)
            if (p !== "default" && !exports.hasOwnProperty(p))
                __createBinding(exports, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    ;
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (Object.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }
    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    function populateInterfaceProperties(target, source, properties) {
        var e_1, _a;
        var allProps = __spread(['name', 'refid', 'title', 'description', 'help'], properties);
        try {
            for (var allProps_1 = __values(allProps), allProps_1_1 = allProps_1.next(); !allProps_1_1.done; allProps_1_1 = allProps_1.next()) {
                var prop = allProps_1_1.value;
                if (source.hasOwnProperty(prop)) {
                    target[prop] = source[prop];
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (allProps_1_1 && !allProps_1_1.done && (_a = allProps_1.return)) _a.call(allProps_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
    }
    var ɵ0 = function (a, b) {
        return 0;
    }, ɵ1 = function (a, b) {
        return a.value.localeCompare(b.value);
    }, ɵ2 = function (a, b) {
        return a.key > b.key ? -1 : (b.key > a.key ? 1 : 0);
    };
    var keyValueSorters = {
        // Preserve original property order
        originalOrder: ɵ0,
        // Order by ascending property value
        valueAscOrder: ɵ1,
        // Order by descending property key
        keyDescOrder: ɵ2
    };

    var FormGroup = /** @class */ (function (_super) {
        __extends(FormGroup, _super);
        function FormGroup(controls) {
            var _this = 
            //super(controls, validatorOrOpts, asyncValidator);
            _super.call(this, controls) || this;
            _this.controls = controls;
            return _this;
        }
        FormGroup.prototype.populate = function (source) {
            populateInterfaceProperties(this, source, []);
        };
        FormGroup.prototype.getControlType = function () {
            return 'group';
        };
        return FormGroup;
    }(forms.FormGroup));

    var FormArray = /** @class */ (function (_super) {
        __extends(FormArray, _super);
        function FormArray(controls, validatorOrOpts, asyncValidator, fbs) {
            var _this = _super.call(this, controls, validatorOrOpts, asyncValidator) || this;
            _this.controls = controls;
            _this.fbs = fbs;
            _this.itemTypes = [];
            _this.additionalItems = false;
            return _this;
        }
        FormArray.prototype.populate = function (source) {
            populateInterfaceProperties(this, source, ['uniqueItems', 'contains', 'additionalItems', 'minItems', 'maxItems']);
        };
        FormArray.prototype.getControlType = function () {
            return 'array';
        };
        FormArray.prototype.setValue = function (value, options) {
            var _this = this;
            if (options === void 0) { options = {}; }
            if (value && value.length > 0) {
                if (!this.fbs)
                    throw 'Must supply FormBuilderService instance.';
                value.forEach(function (val) {
                    var schema = _this.itemTypes.length == 1
                        ? _this.itemTypes[0]
                        : _this.itemTypes.find(function (sch) { return sch.type == typeof val; }); // ToDo: better type matching
                    // console.info("schema+val", schema, val);
                    var ctl = _this.fbs.getFormostAbstractControl(schema);
                    _this.push(ctl);
                });
            }
            _super.prototype.setValue.call(this, value, options);
            // this._checkAllValuesPresent(value);
            // value.forEach((newValue: any, index: number) => {
            //   this._throwIfControlMissing(index);
            //   this.at(index).setValue(newValue, {onlySelf: true, emitEvent: options.emitEvent});
            // });
            // this.updateValueAndValidity(options);
        };
        return FormArray;
    }(forms.FormArray));

    var FormControl = /** @class */ (function (_super) {
        __extends(FormControl, _super);
        function FormControl(source, formState) {
            var _this = _super.call(this, formState, null, null) || this;
            // type: string
            _this.minLength = null;
            _this.maxLength = null;
            _this.pattern = null;
            _this.contentMediaType = null;
            // type: number
            _this.multipleOf = null;
            _this.minimum = null;
            _this.exclusiveMinimum = null;
            _this.maximum = null;
            _this.exclusiveMaximum = null;
            populateInterfaceProperties(_this, source, ['type', 'required', 'enum', 'minLength', 'maxLength', 'pattern', 'format', 'contentEncoding', 'contentMediaType', 'multipleOf', 'minimum', 'exclusiveMinimum', 'maximum', 'exclusiveMaximum']);
            _this.setValidators(_this.getValidators());
            return _this;
        }
        Object.defineProperty(FormControl.prototype, "uiType", {
            get: function () {
                if (this.type == 'string' && this.enum) {
                    return 'dropdown';
                }
                if (this.type == 'boolean') {
                    return 'checkbox';
                }
                // if (this.type == 'string' && (this.minLength > 99 || this.maxLength > 99)) { return 'bigtext'; }
                if (this.type == 'string' && this.contentMediaType == 'text/html') {
                    return 'richtext';
                }
                if (this.type == 'string' && this.contentMediaType == 'text/markdown') {
                    return 'markdown';
                }
                return 'input';
            },
            enumerable: false,
            configurable: true
        });
        ;
        Object.defineProperty(FormControl.prototype, "inputType", {
            get: function () {
                if (this.uiType == 'input') {
                    if (this.type == 'number' || this.type == 'integer') {
                        return 'number';
                    }
                    if (this.type == 'string' && this.format == 'date-time') {
                        return 'datetime-local';
                    }
                    if (this.type == 'string' && this.format == 'date') {
                        return 'date';
                    }
                    if (this.type == 'string' && this.format == 'time') {
                        return 'number';
                    }
                    if (this.type == 'string' && this.format == 'email') {
                        return 'email';
                    }
                    if (this.type == 'string' && this.format == 'uri') {
                        return 'url';
                    }
                    if (this.type == 'string') {
                        return 'text';
                    }
                }
                return null;
            },
            enumerable: false,
            configurable: true
        });
        FormControl.prototype.populate = function (source) {
            populateInterfaceProperties(this, source, ['type', 'required', 'enum', 'minLength', 'maxLength', 'pattern', 'format', 'contentEncoding', 'contentMediaType', 'multipleOf', 'minimum', 'exclusiveMinimum', 'maximum', 'exclusiveMaximum']);
        };
        FormControl.prototype.getControlType = function () {
            return 'control';
        };
        FormControl.prototype.getValidators = function () {
            var validators = [];
            if (this.required)
                validators.push(forms.Validators.required);
            if (this.type == 'string') {
                if (this.minLength !== null)
                    validators.push(forms.Validators.minLength(this.minLength));
                if (this.maxLength !== null)
                    validators.push(forms.Validators.minLength(this.maxLength));
                if (this.pattern !== null)
                    validators.push(forms.Validators.pattern(this.pattern));
                if (this.format == "email")
                    validators.push(forms.Validators.email);
            }
            if (this.type == 'number') {
                if (this.minimum !== null)
                    validators.push(forms.Validators.min(this.minimum));
                if (this.maximum !== null)
                    validators.push(forms.Validators.max(this.maximum));
            }
            console.info(name, validators);
            return validators;
        };
        return FormControl;
    }(forms.FormControl));

    var SchemaConverterService = /** @class */ (function () {
        function SchemaConverterService() {
        }
        SchemaConverterService.prototype.getFormostForm = function (schema) {
            var formItems = this.getFormItems(schema, "#");
            if (formItems instanceof FormGroup)
                return formItems;
            else {
                return new FormGroup({ formost: formItems });
            }
        };
        SchemaConverterService.prototype.getFormostForm$ = function (schema) {
            return rxjs.of(this.getFormostForm(schema));
        };
        SchemaConverterService.prototype.getFormostAbstractControl = function (schema, refId) {
            if (refId === void 0) { refId = "#"; }
            var formItems = this.getFormItems(schema, refId);
            return formItems;
        };
        SchemaConverterService.prototype.getFormostAbstractControl$ = function (schema, refId) {
            if (refId === void 0) { refId = "#"; }
            return rxjs.of(this.getFormostAbstractControl(schema, refId));
        };
        SchemaConverterService.prototype.getFormItems = function (schema, refId) {
            if (refId === void 0) { refId = "#"; }
            var schemaObj = typeof schema === 'string' ? JSON.parse(schema) : schema;
            if (!schemaObj) {
                console.warn('NO-OBJECT', schemaObj);
                return {};
            }
            if (!schemaObj.type) {
                console.warn('NO-TYPE', schemaObj);
                return {};
            }
            var formItems;
            switch (schemaObj.type) {
                case 'null':
                    return null;
                case 'object':
                    formItems = this.getFormGroup(schemaObj, refId);
                    break;
                case 'array':
                    formItems = this.getFormArray(schemaObj, refId);
                    break;
                default:
                    formItems = this.getFormControl(schemaObj, refId);
                    break;
            }
            return formItems;
        };
        SchemaConverterService.prototype.getFormControl = function (obj, refId) {
            var formCtl = new FormControl(obj);
            // formCtl.populate(obj);
            if (refId) {
                formCtl.refid = refId;
            }
            return formCtl;
        };
        SchemaConverterService.prototype.getFormGroup = function (obj, refId) {
            var reqArr = obj.required || [];
            var item = new FormGroup({});
            item.populate(obj);
            if (refId) {
                item.refid = refId;
            }
            for (var prop in obj.properties) {
                if (obj.properties.hasOwnProperty(prop)) {
                    var unknownField = obj.properties[prop];
                    unknownField.name = prop;
                    if (reqArr.indexOf(prop) > -1) {
                        unknownField.required = true;
                    }
                    var knownField = this.getFormItems(unknownField, refId + "/fields/" + prop);
                    //item.fields.push(knownField);
                    item.addControl(prop, knownField);
                }
            }
            return item;
        };
        SchemaConverterService.prototype.getFormArray = function (obj, refId) {
            var item = new FormArray([], undefined, undefined, this);
            item.populate(obj);
            if (refId) {
                item.refid = refId;
            }
            var allowedTypesArray = Array.isArray(obj.items)
                ? obj.items
                : [obj.items];
            // let i = 0;
            // for (const allowedType of allowedTypesArray) {
            //     // ++i;
            //     // allowedType.name = allowedType.name || `${obj.name ? obj.name + '-' : null}type-${i}`;
            //     // item.itemTypes.push(this.getFormItems(allowedType, `${refId}/itemTypes/${allowedType.name}`) as FormField);
            //     item.itemTypes.push(this.getFormItems(allowedType, `${refId}/itemTypes/[${i++}]`) as FormControl);
            // }
            item.itemTypes = allowedTypesArray;
            return item;
        };
        return SchemaConverterService;
    }());
    SchemaConverterService.ɵprov = i0.ɵɵdefineInjectable({ factory: function SchemaConverterService_Factory() { return new SchemaConverterService(); }, token: SchemaConverterService, providedIn: "root" });
    SchemaConverterService.decorators = [
        { type: i0.Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    SchemaConverterService.ctorParameters = function () { return []; };

    /**

    var x: JSONSchema7 = {
        "title": "People",
        "type": "object",
        "properties": {
            "name": {
                "type": "string",
                "minLength": 3,
                "description": "Please enter your name"
            },
            "vegetarian": {
                "type": "boolean",
                "title": "Are you a vegetarian?"
            },
            "birthDate": {
                "type": "string",
                "format": "date"
            },
            "nationality": {
                "type": "string",
                "enum": [
                    "DE",
                    "IT",
                    "JP",
                    "US",
                    "RU",
                    "Other"
                ]
            },
            "favorites": {
                "type": "array",
                "items": {
                    "type": "string"
                },
                "uniqueItems": true,
                "default": []
            },
            "personalData": {
                "type": "object",
                "properties": {
                    "age": {
                        "type": "integer",
                        "description": "Please enter your age."
                    },
                    "height": {
                        "type": "number"
                    },
                    "drivingSkill": {
                        "type": "number",
                        "maximum": 10,
                        "minimum": 1,
                        "default": 7
                    }
                },
                "required": [
                    "age",
                    "height"
                ]
            },
            "occupation": {
                "enum": [
                    "doctor",
                    "lawyer"
                ],
                "type": "string"
            },
            "postalCode": {
                "type": "string",
                "maxLength": 5
            }
        },
        "required": [
            "occupation",
            "nationality"
        ]
    };


    // */

    /*
     * Public API Surface of json-formost-core
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.FormArray = FormArray;
    exports.FormControl = FormControl;
    exports.FormGroup = FormGroup;
    exports.SchemaConverterService = SchemaConverterService;
    exports.keyValueSorters = keyValueSorters;
    exports.populateInterfaceProperties = populateInterfaceProperties;
    exports.ɵ0 = ɵ0;
    exports.ɵ1 = ɵ1;
    exports.ɵ2 = ɵ2;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
